File: 动态规划/64. 最小路径和/最小路径和.cpp
#include<iostream>
#include<vector>
using namespace std;
// 上面的头文件与题解无光,只是为了在本地编译器里不报错

class Solution {
public:
    int minPathSum(vector<vector<int>>& grid) {
        // 在这个问题中,需要注意的就是第一行与第一列的值的路径是唯一确定的,可以先把这些值表示出来
        // 接着由于dp[i][j]的数据来源于上一个数据,也就是左边或者上面的数据的最小的那个值

        // 接下来就是代码
        int height = grid.size();
        int weight = grid[0].size();
        vector<vector<int>> dp(height, vector<int>(weight, 0)); // 初始化dp数组
        dp[0][0] = grid[0][0]; // 初始化起点的节点
        for (int i = 1; i < height; i++) {
            dp[i][0] = dp[i - 1][0] + grid[i][0];
        }
        for (int i = 1; i < weight; i++) {
            dp[0][i] = dp[0][i - 1] + grid[0][i];
        }
        // 下面就是对这个dp进行遍历
        for (int i = 1; i < height; i++) {
            for (int j = 1; j < weight; j++) {
                dp[i][j] = min(dp[i - 1][j], dp[i][j - 1]) + grid[i][j];
            }
        }
        return dp[height - 1][weight - 1];
    }
};

File: 动态规划/64. 最小路径和/最小路径和.go
func minPathSum(grid [][]int) int {
	// 需要注意的是这个题目的移动路径是想左以及向右移动,不能向下移动以及向右移动
	// 这个题目首先对第一行以及第一列进行赋值
	// 接下来就是对除了第一行以及第一列其他的内容进行dp操作
	height, weight := len(grid), len(grid[0])
	dp := make([][]int, height)
	// 初始化这个dp数组
	for i := 0; i < height; i++ {
		dp[i] = make([]int, weight)
	}
	dp[0][0] = grid[0][0]
	for i := 1; i < height; i++ {
		dp[i][0] = dp[i-1][0] + grid[i][0]
	}
	for j := 1; j < weight; j++ {
		dp[0][j] = dp[0][j-1] + grid[0][j]
	}
	for i := 1; i < height; i++ {
		for j := 1; j < weight; j++ {
			dp[i][j] = min(dp[i-1][j], dp[i][j-1]) + grid[i][j]
		}
	}
	return dp[height-1][weight-1]
}